### DISCUSSION TITLE
Maybe make it walk...

### DISCUSSION THREAD

**Discussion by geppetto32** [makerforgetech/modular-biped#69] [Ideas]
Hello everyone... happy new year.... :-) 
first exsuse my english as its not my main ... and second inspired from Dan i desided to maybe 
make the robot walk , sorry  Dan i used other kind of servo and a model that maybe help better on movements..? 
...
Hardware :
https://www.waveshare.com/wiki/ST3215_Servo
https://www.waveshare.com/st3215-hs-servo-motor.htm
https://www.waveshare.com/bus-servo-adapter-a.htm
i used both kind of servos as they serve different speeds  at a given time ....

Software :
ros2 humble
ubuntu 22
pinocchio latest version
gepetto viewer
sdk from waveshare for the servo com
i started from the example atlas, and the pinocchio library ... i twicked a bit the parameters and we have a walking gait generation.... 

https://www.youtube.com/watch?v=dZw76veT1Mk

![a1](https://github.com/makerforgetech/modular-biped/assets/95494535/b09d3349-82c2-43ab-aca0-300ae1a22556)
![a2](https://github.com/makerforgetech/modular-biped/assets/95494535/f03333fd-b479-4f66-b51a-dde0e5b89a58)
![a3](https://github.com/makerforgetech/modular-biped/assets/95494535/df59cb23-6f2e-446f-8452-698e908049a8)

**Comment by WhoIsMrSentry** [comment]
First of all, happy new year! and  wellcome the community :-)

I appreciate your effort in executing your robot. In particular, your approach using two different servos that can operate at different speeds is interesting. The information you shared about hardware and software details is also very useful.

I watched the video explaining the process of moving your robot. It looks pretty impressive! It's nice to see that you're working with the Pinocchio library and ROS2.

I would like to get more details. For example, can you share more information about how you control the speeds of the servos and how you implemented the pinocchio library? Also, can you tell us a little about the challenges you faced while developing your robot and how you overcame those challenges?

@danic85 We are waiting for your comments, maybe we can make it workable with @geppetto32 's contribution.
Well done, you did a great job!

**Comment by geppetto32** [comment]
Yes of  course , i will make a new video, lenghty video i must say. I will include some info besides this particular project with errors and dificulties on how i got here and why. Also i have to mention that i am not a programmer i am a self learning person, learning by mistakes only haha...The video above shows only the sim throught the set of scripts that are neccecery fot that reason. The ros2 implantation is sepperated to make things easier. May the force be with me to include in the video both sim and reality.... talk soon.. :-)

**Comment by danic85** [comment]
Hi @geppetto32 thanks for sharing this, it looks amazing!

I'd be very interested in learning more about the modelling of the robot and training in the simulation. If you can share resources that would be great.

I will say that 12 servos at $30 each makes this design more expensive than a lot of people can afford (myself included, most likely!), but I'm really interested to see where it goes and see if we can make it work for the existing hardware.
The dimensions of the legs are calculated specifically to allow the chosen servos to operate within their torque range, so theoretically an upgrade shouldn't be needed, but there are definitely changes needed to add more actuated DoF to the design.

I'm looking forward to seeing your video when you're ready to share it.

**Comment by geppetto32** [comment]
Hi @danic85 , yes you are correct! its a spicy choice  by adding these servos ....but... my idea is to make a framework so whoever needs to implament his urdf-robot with its parameters,  can easily make the robot walk. I haven't looked if you have a .step file of the companion.So If there is one i could add a hip_roll l,r  joints and have more chances .We need a ZmP" and we can get it only from hip_roll.So as long as i make the video i can modify the .step at the same time (its a matter of minutes).
Below i attach you guys my notes on making the appropriate environment[i dont use env , i install everything globaly(the hard way***wrong thing to do my advise***)]so you can pottentialy see (a-your) robot move :

 You need to install pinocchio, scipy, gepetto-viewer, matplotlib, cvxopt and use python 3.10 in ubuntu 22.
 [
 Before installing, consider that there are other ways to install pinocchio and its despendencies , via conda etc.For more at 2 links below.
https://stack-of-tasks.github.io/pinocchio/download.html 
https://github.com/Gepetto/gepetto-viewer-corba
 its perfered to make a fresh conda env to put everything in there.
]
 
 ............................................ 1 .......................................
  sudo apt install -qqy lsb-release curl
 ......................................................................................
 sudo mkdir -p /etc/apt/keyrings
 curl http://robotpkg.openrobots.org/packages/debian/robotpkg.asc \
     | sudo tee /etc/apt/keyrings/robotpkg.asc    
......................................................................................
     echo "deb [arch=amd64 signed-by=/etc/apt/keyrings/robotpkg.asc] http://robotpkg.openrobots.org/packages/debian/pub $(lsb_release -cs) robotpkg" \
     | sudo tee /etc/apt/sources.list.d/robotpkg.list   
......................................................................................
sudo apt update
......................................................................................
sudo apt install -qqy robotpkg-py3*-pinocchio
......................................................................................
export PATH=/opt/openrobots/bin:$PATH
export PKG_CONFIG_PATH=/opt/openrobots/lib/pkgconfig:$PKG_CONFIG_PATH
export LD_LIBRARY_PATH=/opt/openrobots/lib:$LD_LIBRARY_PATH
export PYTHONPATH=/opt/openrobots/lib/python3.10/site-packages:$PYTHONPATH # Adapt your desired python version here
export CMAKE_PREFIX_PATH=/opt/openrobots:$CMAKE_PREFIX_PATH

may directly add those lines to your $HOME/.bashrc for a persistent configuration. 
gedit ~/.bashrc and paste at the bottom the above 5 exports.    
......................................................................................
 ............................................ 2 ...................................... 
 sudo apt-get install python3-scipy 
 ....................................................................................
 ............................................ 3 .....................................
 sudo apt install robotpkg-pinocchio robotpkg-py310-qt5-gepetto-viewer
  ....................................................................................
 ............................................ 4 .....................................
python -m pip install -U pip
python -m pip install -U matplotlib
....................................................................................
 ............................................ 5 .....................................
 pip install cvxopt
 ....................................................................................
 ............................................ 6 ...................................
 sudo apt update
sudo apt install robotpkg-py310-qt5-gepetto-viewer-corba
....................................................................................

step 3 and 6 are tricky but thats the way i did it and worked.In some cases it could be a bit complex procces ...dont give up ! Its a rewarding task.

Dont forget:  source .bashrc

Hit gepetto-gui and if a window opens with the famous robot ...you are all set.

**Reply by danic85** [reply]
Nice! thanks for the detailed guide :clap:

**Comment by danic85** [comment]
Hi @geppetto32 I realise it has been a while since you posted an update on this, have you made any further progress?
I'm using ST servos now and it would be great to see how far you took this.
